### PR TITLE
Approve Student Exam Accommodations

### PR DESCRIPTION
**Pull request by burzum619**
This pull request includes an endpoint for approving student accommodations, particularly guest/practice test accommodations. 

The new endpoint is found in ExamAccommodationController.approveAccommodations(). The service method called by this method (in ExamAccommodationServiceImpl) is a port of StudentDLL.T_ApproveAccommodations_SP(), which begins at line 11429.

A separate PR will be created for TDS_Student that will cover the Student-related code that will ultimately call this method. The majority of the important logic, however, is contained in this PR.

The "clientname" field has been removed from ApprovalRequest (as it can be obtained by the Exam), which resulted in some changes to miscellaneous unit tests and other test-related code.

### CODE REVIEW

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 1]
```diff
@@ -108,29 +125,72 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+        } else if (maybeSession.isPresent() && maybeSession.get().getProctorId() != null) {
+            return Optional.of(new ValidationError("T_ApproveAccommodations", "Student can only self-approve unproctored sessions"));
+        }
+        
+        // Get the list of current exam accomms in case we need to update them (for example, if a pre-initialized default was changed by the guest user)
+        List<ExamAccommodation> currentAccommodations = examAccommodationQueryRepository.findApprovedAccommodations(examId);
+    
+        // For each assessment and segments separately, initialize their respective accommodations
+        request.getAccommodationCodes().forEach((segmentPosition, guestAccommodationCodes) -> {
+            initializePreviousAccommodations(exam, segmentPosition, false, currentAccommodations, guestAccommodationCodes);
```
Comment (burzum619) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: initializePreviousAccommodations is a port of CommonDll._UpdateOpportunityAccommodations_SP(). Thanks @millerts !

File: service/src/main/java/tds/exam/web/endpoints/ExamAccommodationController.java [hunk 2]
```diff
@@ -1,44 +1,71 @@
...
 
+import static org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo;
+import static org.springframework.hateoas.mvc.ControllerLinkBuilder.methodOn;
+
 @RestController
 @RequestMapping("/exam")
 public class ExamAccommodationController {
     private final ExamAccommodationService examAccommodationService;
-
+
```
Comment (burzum619) [inline service/src/main/java/tds/exam/web/endpoints/ExamAccommodationController.java]: I'm not sure exactly why my auto-formatting is doing this, but unfortunately I didn't catch it till the PR was created. Probably caused by changes I made to my settings for TDS_Student indents/tabs.

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 3]
```diff
@@ -108,29 +125,72 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+    
+    @Override
+    public Optional<ValidationError> approveAccommodations(UUID examId, ApproveAccommodationsRequest request) {
+        /* This method is a port of StudentDLL.T_ApproveAccommodations_SP, starting at line 11429 */
+        ApprovalRequest approvalRequest = new ApprovalRequest(examId, request.getSessionId(), request.getBrowserId());
+    
+        Optional<Exam> maybeExam = examQueryRepository.getExamById(examId);
+        
+        if (!maybeExam.isPresent()) {
+            return Optional.of(new ValidationError("T_ApproveAccommodations", "The test opportunity does not exist"));
```
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: Not sure if this is intentional, but let's move this error code string to the ValdiationErrorCodes constants class and maybe change the "T_..." to something else

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 4]
```diff
@@ -108,29 +125,72 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+        
+        if (maybeError.isPresent()) {
+            return maybeError;
+        }
+    
+        /* lines 11465-11473 */
+        Optional<Session> maybeSession = sessionService.findSessionById(exam.getSessionId());
+        
+        if (!maybeSession.isPresent()) {
+            return Optional.of(new ValidationError("T_ApproveAccommodations", "The test opportunity is not enrolled in this session"));
```
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: The ValidationError codes should be unique, and I think we should put them in the constants file.  Assume the same comment throughout where we're using hte same error code.

File: service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java [hunk 5]
```diff
@@ -108,29 +125,72 @@ public ExamAccommodationServiceImpl(ExamAccommodationQueryRepository examAccommo
...
+    
+        // For each assessment and segments separately, initialize their respective accommodations
+        request.getAccommodationCodes().forEach((segmentPosition, guestAccommodationCodes) -> {
+            initializePreviousAccommodations(exam, segmentPosition, false, currentAccommodations, guestAccommodationCodes);
+        });
+        
+        return Optional.empty();
+    }
+    
+//    private updateExamAccommodations
```
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ExamAccommodationServiceImpl.java]: Remove?

File: service/src/main/java/tds/exam/web/endpoints/ExamAccommodationController.java [hunk 6]
```diff
@@ -1,44 +1,71 @@
...
     @RequestMapping(value = "/{examId}/{segmentId}/accommodations", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
     ResponseEntity<List<ExamAccommodation>> findAccommodations(@PathVariable final UUID examId,
                                                                @PathVariable final String segmentId,
                                                                @RequestParam(name = "type", required = false) final String[] accommodationTypes) {
         return ResponseEntity.ok(examAccommodationService.findAccommodations(examId, segmentId, accommodationTypes));
     }
-
-    @RequestMapping(value = "/{examId}/accommodations")
+    
+    @RequestMapping(value = "/{examId}/accommodations", method = RequestMethod.GET)
```
Comment (millerts) [inline service/src/main/java/tds/exam/web/endpoints/ExamAccommodationController.java]: Good catch

**Pull request merged**